scanner/engine/scheduler.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "scanner/engine/metadata.h"
+#include "scanner/engine/table_meta_cache.h"
+#include "scanner/engine/runtime.h"
+#include "scanner/engine/dag_analysis.h"
+
+#include <deque>
+
+
+namespace scanner {
+namespace internal {
+
+class Scheduler {
+public:
+  // This function should do dependency analysis and fill up global_data_map,
+  // global_dependency_map and global_status_map.
+  void fill_global_maps(DatabaseMetadata& meta, TableMetaCache& table_metas,
+                           const std::vector<proto::Job>& jobs,
+                           const std::vector<proto::Op>& ops,
+                           DAGAnalysisInfo& info);
```

Comment:
It's fairly important that this function can be invoked *incrementally* to allow the runtime to generate only as much work as is needed at a given point in time. The reason for this is simple: if we materialize all the dependencies of every element across, say, 70,000 movies, that can easily blow out the memory of a typical machine. By having the ability to incrementally materialize the dependency structure, we can control how much memory we are using at any given point (it also allows the workers to starts quicker since we can build work in parallel with executing it instead of waiting to create all of it at job start).